File: static/sort/bubble.js
/**
 * Created by YiYing on 2017/4/22.
 */
(function (W) {

    function Bubble(arr) {
        this.arr = arr;
    }

    /**
     * 冒泡排序算法实现
     */
    Bubble.prototype.sort = function () {
        var len = this.arr.length;
        for(var i=0;i<len;i++){
            //将最大的元素不断的沉到底部
            for(var j=0;j<len-i-1;j++){
                if(this.less(j+1,j)){
                    this.exchange(j+1,j);
                }
            }
        }
    };

    /**
     * 判断m是否小于n
     * @param m
     * @param n
     */
    Bubble.prototype.less = function (m,n) {
        //可根据不同的数据类型设置比对规则，比如json。这里适用于数字与字符串。
        return this.arr[m]<this.arr[n];
    };

    /**
     * 交换数组中m与n的位置
     * @param m
     * @param n
     */
    Bubble.prototype.exchange = function (m,n) {
        var swap    = this.arr[m];
        this.arr[m] = this.arr[n];
        this.arr[n] = swap;
    };

    /**
     * 打印排序后的数组
     */
    Bubble.prototype.show = function () {
        console.log(this.arr);
    };

    /**
     * 判断是否已排序
     * @returns {boolean}
     */
    Bubble.prototype.isSorted = function () {
        var len = this.arr.length;
        for(var i=1;i<len;i++){
            if(this.less(i,i-1)){
                return false;
            }
        }
        return true;
    };

    W.Bubble = Bubble;
})(window);

//测试代码
(function () {
    var arr = [4,5,6,0,3,5,21,7,9,0,1];
    var bubble = new Bubble(arr);
    console.log("排序前："+bubble.isSorted());
    bubble.sort();
    console.log("排序后："+bubble.isSorted());
    bubble.show();
})();

File: src/com/algs/sort/Bubble.java
package com.algs.sort;

public class Bubble {
	
	/**
	 * 冒泡排序实现逻辑
	 * @param arr
	 */
	public static void sort(int[] arr){
		int len = arr.length;
		//每遍历一次就把最大的沉到了底部
        for (int i = 0; i < len; i++) {
            for (int j = 0; j < len-i-1; j++) {
            	//如果后面一个元素小于前一个则交换位置
                if (less(arr[j+1], arr[j])){
                	exchange(arr, j+1, j);
                };
            }
        }
	}
	
	/**
	 * 比较m是否小于n
	 * @param m
	 * @param n
	 * @return
	 */
	private static boolean less(int m,int n) {
        return m < n;
    }
	
	/**
	 * 交换m与n的位置
	 * @param a
	 * @param m
	 * @param n
	 */
	private static void exchange(int[] a, int m, int n) {
        int swap = a[m];
        a[m] = a[n];
        a[n] = swap;
    }
	 
	public static void show(int[] a){
		for (int i = 0; i < a.length; i++) {
            System.out.print(a[i]+" ");
        }
	}
	
	public static void main(String[] args) {
		int[] arr = {4,5,6,0,3,5,21,7,9,0,1};
		Bubble.sort(arr);
		Bubble.show(arr);
	}
}
